collection_mutation.cc

```diff
@@ -62,27 +62,47 @@ bool collection_mutation_view::is_empty() const {
 }
 
 bool collection_mutation_view::is_any_live(const abstract_type& type, tombstone tomb, gc_clock::time_point now) const {
-    assert(type.is_collection());
-    auto& ctype = static_cast<const collection_type_impl&>(type);
-
   return data.with_linearized([&] (bytes_view in) {
     auto has_tomb = read_simple<bool>(in);
     if (has_tomb) {
         auto ts = read_simple<api::timestamp_type>(in);
         auto ttl = read_simple<gc_clock::duration::rep>(in);
         tomb.apply(tombstone{ts, gc_clock::time_point(gc_clock::duration(ttl))});
     }
+
     auto nr = read_simple<uint32_t>(in);
-    for (uint32_t i = 0; i != nr; ++i) {
-        auto ksize = read_simple<uint32_t>(in);
-        in.remove_prefix(ksize);
-        auto vsize = read_simple<uint32_t>(in);
-        auto value = atomic_cell_view::from_bytes(ctype.value_comparator()->imr_state().type_info(), read_simple_bytes(in, vsize));
-        if (value.is_live(tomb, now, false)) {
-            return true;
+
+    return visit(type, make_visitor(
+        [&] (const collection_type_impl& ctype) {
+            auto& type_info = ctype.value_comparator()->imr_state().type_info();
+            for (uint32_t i = 0; i != nr; ++i) {
```

Comment:
This loop and the one for UDTs look very similar. It could make sense to generalize them and just pass lambda that reads key and value in collection/udt specific way.